text/0000-demote-apple-32bit.md

```diff
@@ -0,0 +1,125 @@
+`armv7-apple-ios`, `armv7s-apple-ios` and `i386-apple-ios` targets. The release
+after that will demote the targets to Tier 3, which means no official build
+will be available for them, and they will not be tested by CI.
+
+Once this RFC is merged a blog post will be published on the main Rust Blog
+announcing the change, to alert the users of the target of the demotion. The
+demotion will also be mentioned in the release announcement for the last
+release with Tier 1 and Tier 2 support, as well as the first release with Tier
+3 support.
+
+This RFC does **not** propose removing the targets completely from the
+codebase: that will be decided either by another RFC just for those targets, or
+by an RFC defining a general policy for Tier 3 target removal.
+
+Once the targets are demoted to Tier 3, users on other platforms with one of
+those targets' `rust-std` installed won't be able to update the toolchain until
+they remove that target. Users using an Apple 32bit compiler as their host
+platforms will instead be prevented from updating at all, as no new binary
+artifact will be available.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `i686-apple-darwin`, `armv7-apple-ios`, `armv7s-apple-ios` and
+`i386-apple-ios` targets will be considered Tier 3 from the second release
+after this RFC is merged. The code supporting the target will not be removed
+from the compiler, even though we won't guarantee it will continue to work.
+
+The following CI builders will be removed:
+
+- `dist-i686-apple`
+- `i686-apple`
+
+In addition, the `armv7-apple-ios`, `armv7s-apple-ios` and `i386-apple-ios`
+targets will be removed from the `dist-x86_64-apple` builder.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Users might depend on the target, and approving this RFC means they'll be stuck
+on an old compiler version forever, unless they build their own compiler and
+fix the regressions introduced in newer releases themselves.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Support for building 32bit binaries for Apple targets is shrinking as time goes
+by: the latest SDKs from Apple don't support building them at all, and CI
+providers are slowly starting to upgrade their minimum supported SDK versions:
+
+* Azure Pipelines (the provider rustc currently use) doesn't have any public
+  information on when Xcode 9 will be deprecated.
+* GitHub Actions doesn't support Xcode 9 at all.
+* Travis CI [deprecated Xcode older than 9.2][travis-ci-xcode-deprecation] in
+  July 2018.
+
+If this RFC is not accepted, we'll eventually reach a point when we'll have to
+make considerable investments both in terms of money and time to keep building
+on Apple 32bit.
+
+[travis-ci-xcode-deprecation]: https://blog.travis-ci.com/2018-07-19-xcode9-4-default-announce
+
+# Prior art
+[prior-art]: #prior-art
+
+There is no precedent inside the project for the deprecation of Tier 1 targets
```

Comment:
Perhaps the RFC should say more about the general procedure for this sort of thing?

The justification for demotion seems quite reasonable. But it seems unexpected to me that a tier 1 platform would suddenly become tier 3 from one release to another without much warning. I know demotion to tier 3 isn't "removal", but given the less than straightforward process currently of manually building an std targeting a tier 3 platform, it essentially is removal to typical users.

Reply:
> Perhaps the RFC should say more about the general procedure for this sort of thing?

I explicitly didn't want to include a general policy about demoting targets: I hope it will be addressed by https://github.com/rust-lang/rfcs/pull/2803, and I don't want this RFC to get slowed down while we agree on such a policy.

> The justification for demotion seems quite reasonable. But it seems unexpected to me that a tier 1 platform would suddenly become tier 3 from one release to another without much warning. I know demotion to tier 3 isn't "removal", but given the less than straightforward process currently of manually building an std targeting a tier 3 platform, it essentially is removal to typical users.

Unfortunately we don't have much choice on this, sooner or later we're going to have to demote these targets: Apple is actively trying to stop projects from supporting 32bit, as you can't even cross-compile 32bit binaries anymore [starting from macOS 10.14](https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes#3035631).

We don't have any insight on the plans Azure Pipelines has about deprecating build platforms, but we don't expect them to continue providing macOS 10.13 builders forever. I don't expect other CI providers to support that platform after it reaches EOL either.

Once most of the providers offering macOS stop supporting 10.13 we'd then have to build our own build farm, with an out of date, unsupported and potentially vulnerable base OS. That would put even more workload on the already busy infrastructure team, and let's not consider the security risk of doing that.

Unless there are people who actively rely on Apple 32bit support, or other major blockers show up, I don't see why we shouldn't just demote the target soon.

And to be clear, we're happy to support older platforms when it's feasible: all the Linux x86/x86_64 toolchains are built inside a CentOS 5 container to keep supporting Linux 2.6.18, for example. It's just that we don't have the resources to do that if upstream tries to prevent people from doing so.

Reply:
> but given the less than straightforward process currently of manually building an std targeting a tier 3 platform, it essentially is removal to typical users.

Notice that the process is currently quite straightforward: `cargo xbuild --target=i686-apple-darwin` just works and would continue to work as is if this target is demoted to tier 3. We use this in libc to build `libcore` and `libc` for pretty much most/all tier-1, tier-2 and tier-3 targets for which `libc` is a reasonable thing to build (including the `ios` targets mentioned here).

Reply:
It's great if it's that simple. But I though `cargo-xbuild` doesn't support building std? This is mentioned in the [CHANGELOG](https://github.com/rust-osdev/cargo-xbuild/blob/master/CHANGELOG.md) for release v0.4.0 (when it was forked from xargo): "Building std is no longer possible".

If I understand that correctly, and it hasn't changed since then (the README doesn't clearly indicate one way or the other), then I think my statement unfortunately still that it's not currently straightforward to use std on a tier 3 platform.

(Std-aware cargo should greatly improve this.)